logstash-core/src/main/java/org/logstash/instrument/monitors/HotThreadsMonitor.java

```diff
@@ -0,0 +1,191 @@
+                i++;
+            }
+            return sb.toString();
+        }
+
+        Long getWaitedTime() {
+            return (Long)map.get(WAITED_TIME);
+        }
+
+        Long getBlockedTime() {
+            return (Long)map.get(BLOCKED_TIME);
+        }
+
+        Long getCpuTime() {
+            return (Long) map.get(CPU_TIME);
+        }
+    }
+
+    private List<String> VALID_ORDER_BY = new ArrayList<>();
+
+    public HotThreadsMonitor() {
+        VALID_ORDER_BY.add("cpu");
+        VALID_ORDER_BY.add("wait");
+        VALID_ORDER_BY.add("block");
+    }
+
+    /**
+     * Return the current hot threads information as provided by the JVM
+     *
+     * @return A list of ThreadReport including all selected threads
+     */
+    public List<ThreadReport> detect() {
+        Map<String, String> options = new HashMap<String, String>();
+        options.put(ORDERED_BY, "cpu");
+        return detect(options);
+    }
+
+    /**
+     * Return the current hot threads information as provided by the JVM
+     *
+     * @param options Map of options to narrow this method functionality:
+     *                Keys: ordered_by - can be "cpu", "wait" or "block"
+     *                      stacktrace_size - max depth of stack trace
+     * @return A list of ThreadReport including all selected threads
+     */
+    public List<ThreadReport> detect(Map<String, String> options) {
+        String type = "cpu";
+        if (options.containsKey(ORDERED_BY)) {
+            type = options.get(ORDERED_BY);
+            if (!isValidSortOrder(type))
+                throw new IllegalArgumentException("Invalid sort order");
+        }
+
+        int threadInfoMaxDepth = 3;
+        if (options.containsKey(STACKTRACE_SIZE)) {
+            threadInfoMaxDepth = Integer.valueOf(options.get(STACKTRACE_SIZE));
+        }
+
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        enableCpuTime(threadMXBean);
+
+        Map<Long, ThreadReport> reports = new HashMap<>();
+
+        for (long threadId : threadMXBean.getAllThreadIds()) {
+            if (Thread.currentThread().getId() == threadId) {
+                continue;
+            }
+
+            long cpuTime = threadMXBean.getThreadCpuTime(threadId);
+            if (cpuTime == -1) {
+                continue;
+            }
+            ThreadInfo info = threadMXBean.getThreadInfo(threadId, threadInfoMaxDepth);
+            if (info != null) {
+                /*
+                 * Thread ID must exist and be alive, otherwise the threads just
+                 * died in the meanwhile and could be ignored.
+                 */
+                reports.put(threadId, new ThreadReport(info, cpuTime));
+            }
+        }
+        return sort(new ArrayList<>(reports.values()), type);
+     }
+
+    private List<ThreadReport> sort(List<ThreadReport> reports, final String type) {
+        Collections.sort(reports, (a, b) -> {
+           if ("block".equals(type)) {
```

Comment:
Maybe the if block should define the comparator before sort is invoked? 

It's probably not a big deal for performance but aids in readability. 

I believe with JDK 8 there may be a shortcut here for defining a comparator, maybe with a lambda?